### DISCUSSION TITLE
Multiprocessing JSBSim with Python

### DISCUSSION THREAD

**Discussion by sr71684** [JSBSim-Team/jsbsim#516] [Support requests]
When attempting to pass an FGFDMExec instance into a python processing pool, JSBSim returns the following error:
```
  File "stringsource", line 2, in jsbsim.FGFDMExec.__reduce_cython__
TypeError: no default __reduce__ due to non-trivial __cinit__
```
This happens because FGFDMExec can't be trivially pickled. 

Does there exist a way to pickle FGFDMExec such that the above error is avoided? 

**Minimum working example:**

```python
import os
from multiprocessing import Pool

import jsbsim


def make_fdm():
    fdm = jsbsim.FGFDMExec(root_dir=os.path.abspath('/home/jsbsim'))
    fdm.set_debug_level(0)
    time_step = 0
    fdm.load_model("c172p")
    fdm.set_dt(1.0 / 6.0)
    return fdm


def run(fdm):
    # do fdm stuff here
    return fdm

if __name__ == "__main__":
    n_sims = 10
    n_threads = 4
    envs = [make_fdm() for id in range(n_sims)]

    p = Pool(n_threads)
    results = p.map(run, envs)
```

**Environment**
 - OS: Ubuntu 20.04 LTS
 - JSBSim version: v1.1.10
 - Language: Python

**Comment by bcoconni** [comment]
Hi @sr71684,

Indeed [Cython extension types can't be pickled when they contain the `__cinit__` method](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#controlling-pickling).

So your best course of action is to create a new Python class that inherits from `FGFDMExec` and that implements the `__reduce__` method:
```python
class PickleFDMExec(jsbsim.FGFDMExec):
  def load_model(self, model_name):
    self.model_name = model_name  # Store a copy of the model name
    jsbsim.FGFDMExec.load_model(self, model_name)
  def __reduce__(self):
    return rebuild, (self.get_root_dir(), self.model_name, self.get_delta_t())

# Tells pickle how to rebuild the FDM from the reduced data.
def rebuild(root_dir, model_name, dt):
  fdm = jsbsim.FGFDMExec(root=root_dir)
  fdm.load_model(model_name)
  fdm.set_dt(dt)
  return fdm
```

**Reply by sinizu** [reply]
But I get a new problem "__cinit__() got an unexpected keyword argument 'root'", my jsbsim version is 1.1.1. @bcoconni

**Reply by bcoconni** [reply]
Most likely, I got the parameter name `root` wrong. The code should work if you use `root_dir` instead (see the diff below).
```diff
# Tells pickle how to rebuild the FDM from the reduced data.
def rebuild(root_dir, model_name, dt):
-  fdm = jsbsim.FGFDMExec(root=root_dir)
+  fdm = jsbsim.FGFDMExec(root_dir=root_dir)
  fdm.load_model(model_name)
```

**Reply by sinizu** [reply]
```python
import os
from multiprocessing import Pool

import jsbsim
class PickleFDMExec(jsbsim.FGFDMExec):
    def __init__(self) -> None:
        super().__init__()

    def load_model(self, model_name):
        self.model_name = model_name  # Store a copy of the model name
        jsbsim.FGFDMExec.load_model(self, model_name)
    def __reduce__(self):
        return rebuild, (self.get_root_dir(), self.model_name, self.get_delta_t())

def rebuild(root_dir, model_name, dt):
    fdm = jsbsim.FGFDMExec(root_dir=root_dir)
    fdm.load_model(model_name)
    fdm.set_dt(dt)
    return fdm

def make_fdm():
    # fdm = jsbsim.FGFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))
    fdm = PickleFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))
    fdm.set_debug_level(0)
    time_step = 0
    fdm.load_model("c172p")
    fdm.set_dt(1.0 / 6.0)
    return fdm


def run(fdm):
    # do fdm stuff here
    return fdm

if __name__ == "__main__":
    n_sims = 10
    n_threads = 4
    envs = [make_fdm() for id in range(n_sims)]
    p = Pool(n_threads)
    results = p.map(run, envs)
    # make_fdm()
```
_______________________________________________________
The code is as above, but also does not work, "TypeError: __init__() got an unexpected keyword argument 'root_dir'"
:)

**Reply by bcoconni** [reply]
Try removing `root_dir=` or even better switch to the last revision `v1.1.11`.
```diff
def rebuild(root_dir, model_name, dt):
-    fdm = jsbsim.FGFDMExec(root_dir=root_dir)
+    fdm = jsbsim.FGFDMExec(root_dir)
    fdm.load_model(model_name)
```

**Reply by seanmcleod** [reply]
>  "TypeError: init() got an unexpected keyword argument 'root_dir'"

You've added an `__init__(self)` method to the `PickleFDMExec` class.

But you're passing a named argument `root_dir` via
`fdm = PickleFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))`

But your `__init__` method doesn't handle any named arguments. Isn't that what the `TypeError` is trying to tell you?

**Reply by bcoconni** [reply]
@seanmcleod Good catch ! :+1:

**Reply by sinizu** [reply]
```python
import os
from multiprocessing import Pool

import jsbsim
class PickleFDMExec(jsbsim.FGFDMExec):
    def load_model(self, model_name):
        self.model_name = model_name  # Store a copy of the model name
        jsbsim.FGFDMExec.load_model(self, model_name)
    def __reduce__(self):
        return rebuild, (self.get_root_dir(), self.model_name, self.get_delta_t())

def rebuild(root_dir, model_name, dt):
    fdm = jsbsim.FGFDMExec(root_dir)
    fdm.load_model(model_name)
    fdm.set_dt(dt)
    return fdm

def make_fdm():
    # fdm = jsbsim.FGFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))
    fdm = PickleFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))
    fdm.set_debug_level(0)
    fdm.load_model("c172p")
    fdm.set_dt(1.0 / 6.0)
    return fdm


def run(fdm):
    # do fdm stuff here
    return fdm

if __name__ == "__main__":
    n_sims = 10
    n_threads = 4
    envs = [make_fdm() for id in range(n_sims)]
    p = Pool(n_threads)
    results = p.map(run, envs)
    # make_fdm()
```
_____________________________________________________________

```bash
     JSBSim Flight Dynamics Model v1.1.11 [GitHub build 741/commit 10e077b486c42d38e76d1dd7f0aeb744f92d96a0] Feb 13 2022 12:40:33
            [JSBSim-ML v2.0]

JSBSim startup beginning ...

Traceback (most recent call last):
  File "/home/sinizu/code/RL/MAPPO/JSBsim/MAPPO_my_v6_jsbsim_without_state_trans/envs/test.py", line 37, in <module>
    results = p.map(run, envs)
  File "/home/sinizu/anaconda3/envs/marl_36/lib/python3.6/multiprocessing/pool.py", line 266, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/sinizu/anaconda3/envs/marl_36/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
multiprocessing.pool.MaybeEncodingError: Error sending result: '[FGFDMExec 
root dir        :       /home/sinizu/anaconda3/envs/marl_36/share/JSBSim
aircraft path   :       /home/sinizu/anaconda3/envs/marl_36/share/JSBSim/aircraft
engine path     :       /home/sinizu/anaconda3/envs/marl_36/share/JSBSim/engine
systems path    :       /home/sinizu/anaconda3/envs/marl_36/share/JSBSim/systems
output path     :       /home/sinizu/anaconda3/envs/marl_36/share/JSBSim/.
]'. Reason: 'TypeError('no default __reduce__ due to non-trivial __cinit__',)'
```

**Reply by sinizu** [reply]
> > "TypeError: init() got an unexpected keyword argument 'root_dir'"
> 
> You've added an `__init__(self)` method to the `PickleFDMExec` class.
> 
> But you're passing a named argument `root_dir` via `fdm = PickleFDMExec(root_dir=os.path.abspath('/home/sinizu/anaconda3/envs/marl_36/share/JSBSim'))`
> 
> But your `__init__` method doesn't handle any named arguments. Isn't that what the `TypeError` is trying to tell you?

Thank you. That solves part of the problem

**Reply by seanmcleod** [reply]
I've never done any pickling myself, so I'm sort of trying to make educated guesses.

So it looks like it's not seeing the implemented `__reduce__` method you added to the wrapper class `PickleFDMExec`, maybe the following:

```python
def rebuild(root_dir, model_name, dt):
    fdm = jsbsim.FGFDMExec(root_dir)
```

Should be changed to return the wrapper class object?

```python
def rebuild(root_dir, model_name, dt):
    fdm = PickleFDMExec(root_dir)
```

**Reply by sinizu** [reply]
> 

work perfectly!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Thanks a lot!! mua :)

**Reply by sinizu** [reply]
> Try removing `root_dir=` or even better switch to the last revision `v1.1.11`.
> 
> ```diff
> def rebuild(root_dir, model_name, dt):
> -    fdm = jsbsim.FGFDMExec(root_dir=root_dir)
> +    fdm = jsbsim.FGFDMExec(root_dir)
>     fdm.load_model(model_name)
> ```

Also thank you for your advice!

**Reply by seanmcleod** [reply]
Great. Out of interest what are you doing with the multiple JSBSim instances in the thread pool? Is this for some sort of reinforcement learning?

Just an fyi, in terms of multiple instances of JSBSim running in a thread pool you can thank @bcoconni for enabling that a while back by modifying some code in JSBSim that used to maintain some static variable(s) which would cause havoc/crashes when used by multiple threads.

**Reply by bcoconni** [reply]
> I've never done any pickling myself, so I'm sort of trying to make educated guesses.
> 
> So it looks like it's not seeing the implemented `__reduce__` method you added to the wrapper class `PickleFDMExec`, maybe the following:
> 
> ```python
> def rebuild(root_dir, model_name, dt):
>     fdm = jsbsim.FGFDMExec(root_dir)
> ```
> 
> Should be changed to return the wrapper class object?
> 
> ```python
> def rebuild(root_dir, model_name, dt):
>     fdm = PickleFDMExec(root_dir)
> ```

Good to know. This requirement to return a `PickleFDMExec` instance seems to be specific to the `multiprocessing` library because the following test is working on my platform:
```python
import jsbsim
import pickle

class PickleFGFDMExec(jsbsim.FGFDMExec):
     def load_model(self, model_name):
         self.model_name = model_name
         jsbsim.FGFDMExec.load_model(self, model_name)
     def __reduce__(self):
         return rebuild, (self.get_root_dir(), self.model_name, self.get_delta_t())

def rebuild(root_dir, model_name, dt):
     fdm = jsbsim.FGFDMExec(root_dir)
     fdm.load_model(model_name)
     fdm.set_dt(dt)
     return fdm

fdm = PickleFGFDMExec(None)  # Only works on JSBSim v1.1.10 and higher
fdm.load_model("c172p")

with open('fdm.pickle', 'wb') as f:
    pickle.dump(fdm, f, pickle.HIGHEST_PROTOCOL)

with open('fdm.pickle', 'rb') as f:
    fdm2 = pickle.load(f)

fdm2.run_ic()
```

Note that `rebuild` is returning an `FGFDMExec` instance instead of a `PickleFGFDMExec` instance.

**Reply by sinizu** [reply]
> Great. Out of interest what are you doing with the multiple JSBSim instances in the thread pool? Is this for some sort of reinforcement learning?
> 
> Just an fyi, in terms of multiple instances of JSBSim running in a thread pool you can thank @bcoconni for enabling that a while back by modifying some code in JSBSim that used to maintain some static variable(s) which would cause havoc/crashes when used by multiple threads.

Yes, I use multiple processes to speed up the interaction of agents with the environment.

**Reply by seanmcleod** [reply]
@bcoconni so `type(fdm) == PickleFGFDMExec` but `type(fdm2) == FGFDMExec` right?

So if you now tried to pickle `fdm2` via `pickle.dump()` I'm guessing you'll run into the issue @sinizu was seeing with multiprocessing. My guess is multiprocessing is trying re-prickle instances at some point.

**Reply by bcoconni** [reply]
> @bcoconni so `type(fdm) == PickleFGFDMExec` but `type(fdm2) == FGFDMExec` right?

Correct.
> So if you now tried to pickle `fdm2` via `pickle.dump()` I'm guessing you'll run into the issue @sinizu was seeing with multiprocessing.

Correct.
> My guess is multiprocessing is trying re-prickle instances at some point.

Most likely yes. There are a number of places in [the docs of `multiprocessing`](https://docs.python.org/3/library/multiprocessing.html) that require that objects must be "picklable". I was not expecting that unpickled objects could need to be re-pickled afterward so my test case was not representative of @sinizu's request.

**Comment by bcoconni** [comment]
This topic has been moved to JSBSim Discussions as per the issue template statement: `Issues that are neither bug reports or feature requests will be transferred to [Discussions](https://github.com/JSBSim-Team/jsbsim/discussions) by the team and the issue will be locked and closed.`.

Support for this topic will therefore continue within this discussion. Thanks for your understanding.
